File: 49-group-anagrams/group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        dictionary = dict()
        for word in strs:
            key = "".join(sorted(word))   # hashable, and sorted() returns a value
            if key not in dictionary:
                dictionary[key] = [word]  # list of originals, not list(word)
            else:
                dictionary[key].append(word)
        return list(dictionary.values())

File: 49-group-anagrams/group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String,List<String>> groupToWords = new HashMap<>();
        for( String word: strs){
            char[] chars = word.toCharArray();
            Arrays.sort(chars);
            String key = new String(chars);
            groupToWords.computeIfAbsent(key, k -> new ArrayList<>()).add(word);
        }

        return new ArrayList<>(groupToWords.values());
    }
}